Check for negative "from" values in search request body

cbuescher opened the pull request:
Today we already disallow negative values for the "from" parameter in the search
API when it is set as a request parameter and setting it on the
SearchSourceBuilder, but it is still parsed without complaint from a search
body, leading to different, confusing exceptions later on. This PR changes 
this behavior to be the same regardless of setting the value directly, as url 
parameter or in the search body. Values of -1 are still allowed as the current 
default, meaning an unset value, in order not to break existing searches and 
templates using this.

Closes #54897

Commit:
Forbid negative "from" in search request body

Today we already disallow negative values for the "from" parameter in the search
API when it is set as a request parameter and setting it on the
SearchSourceBuilder, but it is still parsed without complaint from a search
body, leading to differing exceptions later. This PR changes this behavior to be
the same regardless of setting the value directly, as url parameter or in the
search body. Values of -1 are still allowed as the current default, meaning an
unset value, in order not to break existing searches and templates using this.

Closes #54897

elasticmachine commented:
Pinging @elastic/es-search (:Search/Search)

Commit:
checkstyle

Commit:
skip on mixed cluster for now

cbuescher commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix-54897

cbuescher commented:
@elasticmachine elasticsearch-ci/2

cbuescher commented:
@elasticmachine run elasticsearch-ci/2

jimczi commented on rest-api-spec/src/main/resources/rest-api-spec/test/search/30_limits.yml:
```diff
@@ -26,14 +26,49 @@ setup:
+      search:
+        rest_total_hits_as_int: true
+        index: test_1
+        # value of -1 should be accepted as unset value
```
Can we try to address this ? For instance we could set the default value directly rather than delaying this to the creation of the search context ?

cbuescher replied:
I'd be open to forbid users setting from to  "-1" entirely on 8.0 but since we currently accept it in the request body (and treat it as 0 later) I think we should allow it at least on 7.x to not break existing queries there. Every other negative value should alread throw a different exception, so thats fine I think. We can log a warning on 7.x when using "-1" to prepare for the change in 8.0 if thats okay?

jimczi commented on server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -336,10 +336,13 @@ public QueryBuilder postFilter() {
     public SearchSourceBuilder from(int from) {
-        if (from < 0) {
-            throw new IllegalArgumentException("[from] parameter cannot be negative");
+        // accept -1 as the default unset value without error
```
Can we set the default early so that we can refuse any negative values ?

cbuescher replied:
I think for `SearchSourceBuilder` I would prefer to keep the notion of an "unset" value and interpret that in any way necessary later. I think this class should represent the user request body, e.g. in `innerToXContent` we only output values that are explicitely set by the user. Maybe @javanna has some preferences there as well?

jimczi reviewed (changes requested):
I left one comment.

jtibshirani commented on docs/reference/search/request/from-size.asciidoc:
```diff
@@ -6,9 +6,9 @@ parameters. The `from` parameter defines the offset from the first
-defaults to `10`.
+Though `from` and `size` can be set as request parameters, they can also be set
+within the search body. `from` defaults to `0`, and `size` defaults to `10`.
+Both values cannot be negative.
```
Small comment, I think 'Neither value can be negative.' would be clearer.

Commit:
change wording

cbuescher commented:
@jimczi I left some answers to the questions you raised in your review. I'm leaning towards forbidding users setting `from` to "-1" entirely on 8.0, I think in 7.x we cannot do that since it would be breaking but log a warning instead. I currently would prefer to keep the distinction between an "unset" value for the reasons stated above but am open to discussion, maybe also involving input from @javanna

Commit:
Merge branch 'master' into fix-54897

cbuescher commented:
@jimczi @javanna just found this PR sitting around, let me know if there anything I can do to move it forward.

Commit:
Make -1 an invalid  parameter setting

cbuescher commented:
@jimczi thanks for the review, I updated the PR to reject `from` values of -1 as url and search body parameter, however I still think this should be considered a breaking change then so I added a note to the migration docs and would only merge to master at this point.

cbuescher commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix-54897

jimczi reviewed (approved):
LGTM
